test/extended/etcd/vertical_scaling.go

```diff
@@ -112,4 +112,85 @@ var _ = g.Describe("[sig-etcd][Serial] etcd [apigroup:config.openshift.io]", fun
 		err = errors.Wrap(err, "scale-down: timed out waiting for only 3 Running master machines")
 		o.Expect(err).ToNot(o.HaveOccurred())
 	})
+
+	// The following test covers replacing a failed master machine from 3 master nodes cluster.
+	// It starts by setting one of the master machine to be failed/unhealthy.
+	// After the failed machine is being deleted, it is expected to be replaced automatically
+	// The test ends validating the size of the cluster, and asserting the member was removed from the etcd cluster by contacting MemberList API.
+	g.It("is able to replace failed master machine from 3 master nodes cluster [Timeout:60m][apigroup:machine.openshift.io]", func() {
+		// set up
+		ctx := context.TODO()
+		etcdClientFactory := scalingtestinglibrary.NewEtcdClientFactory(oc.KubeClient())
+		machineClientSet, err := machineclient.NewForConfig(oc.KubeFramework().ClientConfig())
+		o.Expect(err).ToNot(o.HaveOccurred())
+		machineClient := machineClientSet.MachineV1beta1().Machines("openshift-machine-api")
+		kubeClient := oc.KubeClient()
+
+		// make sure it can be run on the current platform
+		scalingtestinglibrary.SkipIfUnsupportedPlatform(ctx, oc)
+
+		// assert the cluster state before we run the test
+		err = scalingtestinglibrary.EnsureInitialClusterState(ctx, g.GinkgoT(), etcdClientFactory, machineClient, kubeClient)
+		err = errors.Wrap(err, "pre-test: timed out waiting for initial cluster state to have 3 running machines and 3 voting members")
+		o.Expect(err).ToNot(o.HaveOccurred())
+
+		// step 0: ensure clean state after the test
+		defer func() {
+			// since the deletion triggers a new rollout
+			// we need to make sure that the API is stable after the test
+			// so that other e2e test won't hit an API that undergoes a termination (write request might fail)
+			g.GinkgoT().Log("cleaning routine: ensuring initial cluster state and waiting for api servers to stabilize on the same revision")
+			err = scalingtestinglibrary.EnsureInitialClusterState(ctx, g.GinkgoT(), etcdClientFactory, machineClient, kubeClient)
+			err = errors.Wrap(err, "cleaning routine: timed out while ensuring cluster state back to 3 running machines and 3 voting members")
+			o.Expect(err).ToNot(o.HaveOccurred())
+
+			err = testlibraryapi.WaitForAPIServerToStabilizeOnTheSameRevision(g.GinkgoT(), oc.KubeClient().CoreV1().Pods("openshift-kube-apiserver"))
+			err = errors.Wrap(err, "cleaning routine: timed out waiting for APIServer pods to stabilize on the same revision")
+			o.Expect(err).ToNot(o.HaveOccurred())
+		}()
+
+		// step 1: terminate ec2 instance backing a random master machine, resulting in unhealthy etcd cluster
+		victimMachineName, err := scalingtestinglibrary.FailStopRandomMasterMachine(ctx, g.GinkgoT(), kubeClient, machineClient)
+		o.Expect(err).ToNot(o.HaveOccurred())
+
+		// recreate etcd client to existing master node
+		etcdClientFactory = scalingtestinglibrary.NewEtcdClientFactory(oc.KubeClient())
```

Comment:
@tjungblu  would this be sufficient to recreate the `etcdclient` to another existing member ? 

I am just trying to reuse the helpers

Reply:
I don't think this makes a difference because the issue stems from the way the etcd service works with http2:

https://github.com/openshift/origin/blob/0cfbdcd28cf098e3d79e628d2e5e1b42b2a03dea/test/extended/etcd/helpers/client.go#L33-L35

But maybe we're lucky and your test goes green anyway.

Reply:
I am running it now against a cluster and it seems better .. 

Yes i saw it uses the service, iiuc it will use on of the endpoints in the service .

but what is the correct way to have deterministic behaviour ?

Reply:
the cluster recovered and terminated member removed, a replacement was created and added 

the e2e itself still stuck .. the problem in the client, i am stuck with it :/

Reply:
Ideally you would pick a master node that is not the one you intend to kill, I assume you would need to change how `EnsureVotingMembersCount` creates and uses the proxy. 

It does seem like it creates a new client all the time, maybe there's a port clash, do you see that in your logs?

Reply:
here are all the logs https://github.com/openshift/origin/pull/27496#issuecomment-1297183308 .. tbh i dont know what is the issue :/ .. do u have few minutes ?

Reply:
it did, a new machine came was created as replacement and joined the etcd cluster .. i checked the logs myself

Reply:
**etcdctl output**

```
sh-4.4# etcdctl endpoint status -w table
{"level":"warn","ts":"2022-10-31T14:17:52.205Z","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc00045e000/10.0.151.23:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = context deadline exceeded"}
Failed to get the status of endpoint https://10.0.189.153:2379 (context deadline exceeded)
+--------------------------+------------------+---------+---------+-----------+------------+-----------+------------+--------------------+--------+
|         ENDPOINT         |        ID        | VERSION | DB SIZE | IS LEADER | IS LEARNER | RAFT TERM | RAFT INDEX | RAFT APPLIED INDEX | ERRORS |
+--------------------------+------------------+---------+---------+-----------+------------+-----------+------------+--------------------+--------+
| https://10.0.151.23:2379 | 7d59549bb5563cbc |   3.5.5 |   84 MB |     false |      false |         8 |      67569 |              67569 |        |
| https://10.0.230.36:2379 | a623bf12e6883a04 |   3.5.5 |   84 MB |      true |      false |         8 |      67596 |              67596 |        |
+--------------------------+------------------+---------+---------+-----------+------------+-----------+------------+--------------------+--------+
sh-4.4# etcdctl member list -w table
+------------------+---------+-------------------------------------------+--------------------------+--------------------------+------------+
|        ID        | STATUS  |                   NAME                    |        PEER ADDRS        |       CLIENT ADDRS       | IS LEARNER |
+------------------+---------+-------------------------------------------+--------------------------+--------------------------+------------+
| 19c5f03cef63dcb0 | started | ip-10-0-185-41.us-west-1.compute.internal | https://10.0.185.41:2380 | https://10.0.185.41:2379 |      false |
| 7d59549bb5563cbc | started | ip-10-0-151-23.us-west-1.compute.internal | https://10.0.151.23:2380 | https://10.0.151.23:2379 |      false |
| a623bf12e6883a04 | started | ip-10-0-230-36.us-west-1.compute.internal | https://10.0.230.36:2380 | https://10.0.230.36:2379 |      false |
+------------------+---------+-------------------------------------------+--------------------------+--------------------------+------------+
sh-4.4# command terminated with exit code 137
~/workspace/go/src/tmp ᐅ oc rsh -n openshift-etcd pod/etcd-ip-10-0-230-36.us-west-1.compute.internal
sh-4.4# etcdctl member list -w table
+------------------+---------+-------------------------------------------+--------------------------+--------------------------+------------+
|        ID        | STATUS  |                   NAME                    |        PEER ADDRS        |       CLIENT ADDRS       | IS LEARNER |
+------------------+---------+-------------------------------------------+--------------------------+--------------------------+------------+
| 19c5f03cef63dcb0 | started | ip-10-0-185-41.us-west-1.compute.internal | https://10.0.185.41:2380 | https://10.0.185.41:2379 |      false |
| 7d59549bb5563cbc | started | ip-10-0-151-23.us-west-1.compute.internal | https://10.0.151.23:2380 | https://10.0.151.23:2379 |      false |
| a623bf12e6883a04 | started | ip-10-0-230-36.us-west-1.compute.internal | https://10.0.230.36:2380 | https://10.0.230.36:2379 |      false |
+------------------+---------+-------------------------------------------+--------------------------+--------------------------+------------+
sh-4.4# etcdctl endpoint status -w table
+--------------------------+------------------+---------+---------+-----------+------------+-----------+------------+--------------------+--------+
|         ENDPOINT         |        ID        | VERSION | DB SIZE | IS LEADER | IS LEARNER | RAFT TERM | RAFT INDEX | RAFT APPLIED INDEX | ERRORS |
+--------------------------+------------------+---------+---------+-----------+------------+-----------+------------+--------------------+--------+
| https://10.0.151.23:2379 | 7d59549bb5563cbc |   3.5.5 |   94 MB |      true |      false |        10 |      70143 |              70143 |        |
| https://10.0.185.41:2379 | 19c5f03cef63dcb0 |   3.5.5 |   93 MB |     false |      false |        10 |      70143 |              70143 |        |
| https://10.0.230.36:2379 | a623bf12e6883a04 |   3.5.5 |   94 MB |     false |      false |        10 |      70143 |              70143 |        |
+--------------------------+------------------+---------+---------+-----------+------------+-----------+------------+--------------------+--------+
sh-4.4# etcdctl member list -w table
```

**etcdoperator logs**

```
I1031 14:14:58.610248       1 event.go:285] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"openshift-etcd-operator", Name:"etcd-operator", UID:"31143258-bb3e-4514-932c-f25ac094c289", APIVersion:"apps/v1", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'MemberAddAsLearner' successfully added new member https://10.0.185.41:2380
I1031 14:14:58.634180       1 clustermembercontroller.go:283] Not ready for promotion: etcd learner member (https://10.0.185.41:2380) is not yet in sync with leader's log 
I1031 14:15:03.910152       1 event.go:285] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"openshift-etcd-operator", Name:"etcd-operator", UID:"31143258-bb3e-4514-932c-f25ac094c289", APIVersion:"apps/v1", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'MemberPromote' successfully promoted learner member https://10.0.185.41:2380
```

Reply:
u see in the logs and the tables the added member and the corresponding events/logs :)

Reply:
oh now I see, it reports it here:

```
**cluster has reached the expected number of 3 voting members, the members are: [ip-10-0-185-41.us-west-1.compute.internal ip-10-0-151-23.us-west-1.compute.internal ip-10-0-230-36.us-west-1.compute.internal]**


cleaning routine: ensuring initial cluster state and waiting for api servers to stabilize on the same revision
Trying up to 5m0s to recover the cluster to its initial state
W1031 15:15:06.832974   61166 clientconn.go:1331] [core] grpc: addrConn.createTransport failed to connect to {127.0.0.1:63861 127.0.0.1 <nil> 0 <nil>}. Err: connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:63861: operation was canceled". Reconnecting...
checking if there are any excessive machines in the cluster (created by a previous test), expected cluster size is 3, found 3 machines: [ci-ln-dm8bch2-76ef8-9f9pl-master-1 ci-ln-dm8bch2-76ef8-9f9pl-master-2 ci-ln-dm8bch2-76ef8-9f9pl-master-s7pwq-0]
Waiting up to 25m0s for the cluster to reach the expected member count of 3
cluster has reached the expected number of 3 voting members, the members are: [ip-10-0-185-41.us-west-1.compute.internal ip-10-0-151-23.us-west-1.compute.internal ip-10-0-230-36.us-west-1.compute.internal]
Waiting up to 10m0s for the cluster to reach the expected machines count of 3
W1031 15:15:40.514644   61166 clientconn.go:1331] [core] grpc: addrConn.createTransport failed to connect to {127.0.0.1:63872 127.0.0.1 <nil> 0 <nil>}. Err: connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:63872: operation was canceled". Reconnecting...
[AfterEach] [sig-etcd][Serial] etcd [apigroup:config.openshift.io]
  github.com/openshift/origin/test/extended/util/client.go:158
[AfterEach] [sig-etcd][Serial] etcd [apigroup:config.openshift.io]
  github.com/openshift/origin/test/extended/util/client.go:159
[AfterEach] [sig-etcd][Serial] etcd [apigroup:config.openshift.io]
  github.com/openshift/origin/test/extended/etcd/vertical_scaling.go:30
```

the issue is somewhere in the cleanup

Reply:
hmmmm .. i see .. good catch 👍🏽